shoop/core/models/addresses.py

```diff
@@ -173,6 +117,82 @@ def as_string_list(self, locale=None):
     def __iter__(self):
         return iter(self.as_string_list())
 
+    def to_immutable(self):
+        """
+        Get or create saved ImmutableAddress from self.
+
+        :rtype: ImmutableAddress
+        :return: Saved ImmutableAddress with same data as self.
+        """
+        data = get_data_dict(self)
+        return ImmutableAddress.from_data(data)
+
+    def to_mutable(self):
+        """
+        Get a new MutableAddress from self.
+
+        :rtype: MutableAddress
+        :return: Fresh unsaved MutableAddress with same data as self.
+        """
+        data = get_data_dict(self)
+        return MutableAddress.from_data(data)
+
+
+class MutableAddress(Address):
+    """
+    An address that can be changed.
+
+    Mutable addresses are used for e.g. contact's saved addresses.
+    They are saved as new immutable addresses when used in e.g. orders.
+
+    Mutable addresses can be created with `MutableAddress.from_data`
+    or with the `to_mutable` method of `Address` objects.
+    """
+    @classmethod
+    def from_data(cls, data):
+        """
+        Construct mutable address from a data dictionary.
+
+        :param data: data for address
+        :type data: dict[str,str]
+        :return: Unsaved mutable address
+        :rtype: MutableAddress
+        """
+        return cls(**data)
+
+
+class ImmutableAddress(ChangeProtected, Address):
+    """
+    An address that can not be changed.
+
+    Immutable addresses are used for orders, etc., where subsequent
+    edits to the original address (for example an user's default address)
+    must not affect past business data.
+
+    Immutable addresses can be created directly, with the `.from_data()`
+    method, or by creating an immutable copy of an existing `MutableAddress`
+    with the `Address.to_immutable()` method.
+    """
+    @classmethod
+    def from_data(cls, data):
+        """
+        Create immutable address from a data dictionary.
```

Comment:
This should have a mention of the fact that this might not create a new address.